### PR TITLE
fix: default cors, add * as an allow-all configuration, don't add localhost to allowed cors hosts

### LINKED ISSUE #1640: Missing CORS Response Header
When making calls from a browser that includes CORS request header Origin, the server does not include correct CORS response header Access-Control-Allow-Origin.

The call looks fine from the server. It’s the browser that blocks the response because of it’s Same Origine Policy. This policy is in effet in all modern browsers and only enforced on cross-site requests. It forces the browser to include a Origin request header, and blocks all request responses where header Access-Control-Allow-Origin is missing, or where the response header is present but does not list the requesting origin (or * to allow any origin).

I've looked at the SK's Security/Allowed CORS Origins and tried values: * and other URLs, but the header is never present in the call responses.

This could also be true for all SK services, not sure... I remember Login and logout had this problem a while ago... 

This can be reproduced using a browser app running on one server and making calls to a remote Signal K server. Such as using KIP from you local Pi installation and using demo.singalk.org as the stream server or, if you code and debug from a computer that make stream connection to a remote Signal K server (Mac Dev laptop -> remote Signal K Pi box).

You will not see this error if you run the app on the SK box that you stream from because it's simply not a Cross Site Scripting scenario and so the bowser's CORS Same Origine Policy is not applied, or if it's not a browser Webapp (obviously).

image

image

image

Thanks

### PR DESCRIPTION
**Pull request by tkurki**
- Fixes #1640 : #1581 accidentally disabled the default CORS behavior, that allows all CORS origins with `Access-Control-Allow-Origin: *`
- fix: do not add `localhost` addresses to the CORS origins when saved from the Admin UI
- feature: add a new special `*` value for allowed CORS origins. This will allow all origins and echo back the origin in `Access-Control-Allow-Origin` instead of the default *. This allows configuring the server to work with security enabled and `credentials:include` requests from all origins

### CODE REVIEW

**Comment by tkurki** [comment]
@JacoKoster feel like giving this a go, as in "could you please help test this"? The same goes for @godind

File: src/cors.ts [hunk 1]
```diff
@@ -9,15 +9,29 @@ export function setupCors(
...
-  const corsOptions: CorsOptions = {
-    credentials: true,
-    origin: corsOrigins
+
+  // default wildcard cors configuration does not work
+  // with credentials:include client requests, so add
+  // our own wildcard rule that will match all origins
+  // but respond with that origin, not the default *
+  if (allowedCorsOrigins?.startsWith('*')) {
+    corsOptions.origin = (origin: string | undefined, cb) => cb(null, origin)
```
Comment (JacoKoster) [inline src/cors.ts:27]: Wouldn't it be better instead of doing this to set corsOptions.credentials to false?

Comment (tkurki) [reply src/cors.ts:27]: The use case is cors requests with credentials sent by the client - supporting authenticated cors requests. Without credentials true that won’t work.

Comment (JacoKoster) [reply src/cors.ts:27]: Hmm, ok. We are basically solving the client-issue by allowing anything that comes in. Feels odd, but understand the solution.

**Comment by JacoKoster** [comment]
Just tested it and looks good 👍🏼

File: src/cors.ts (hunk 1 above)
Comment (tkurki) [reply src/cors.ts:27]: Only if you enter *. You can still limit by origin or disallow credentials by default config with empty allowedOrigins.

**Comment by godind** [comment]
Made some tests and I think we still have a couple of issues.

- Problem 1: The SK server's CORS implementation should not block any incoming requests based on Allowed CORS Origins configuration value. I've not checked the code...but I this was implemented - see Problem 2.
Explanation: CORS is a Browser security policy. Its intent is for the browser to prevent sharing the server response to its front-end (user, script, processes etc.) for illegitimate hijacked cross-site requests. Look at the call result details, even when you have a CORS error in the browser, the call status is actually OK 200; meaning the request was processed properly, but the browser is not allowing the response to propagate.

- Problem 2: When using the admin UI to edit Allowed CORS Origins and add `*`, it should not automatically add the local hostname. It's never required. I guess it's been done this way because because of Problem 1. If you manually edit security.json and only keep `*` the Admin UI stop working and it should not. CORS does not apply to same-site requests (Request header `Sec-Fetch-Site: same-origin`) and so it should not be a problem when you run Admin UI from the server to access services of that same server.

> NOTE: Chromium does not apply/include CORS security so none of the CORS issues are visible. Use Chrome, Safari, etc. to test.

- Problem 3: I think the config options should be as follows for `allowedCorsOrigins`:
1. No value `""` = CORS not enable.
2. `*` = all hostnames enabled
3. list of hosts = only those hostnames listed.
4. We should not mix a hostname list and `*`. It's one or the other.

> If you want to enable CORS for every hostname at install: have `allowedCorsOrigins: *` as the default config.

Here is the behaviour I think we need depending on values of `allowedCorsOrigins`:

1. "" - return `Access-Control-Allow-Credential: true` & no `Access-Control-Allow-Origin` response header.
2. "*" - return `Access-Control-Allow-Credential: true` & `Access-Control-Allow-Origin: <copy of Request Header's Origin value>`
3. "list of hostname(s)" - return `Access-Control-Allow-Credential: true` & `Access-Control-Allow-Origin:` if Request Header's Origin is found in the list, return it. Else no `Access-Control-Allow-Origin` header

Hope this helps. Long test period!

**Comment by godind** [comment]
Sorry if you received an email when I first commented... I've since made edits. Please read the latest.

**Comment by tkurki** [comment]
@godind what did you test? I think all your points are already in this PR, except empty value disabling CORS. I see very little value in that and it would change the current default behavior.

**Comment by JacoKoster** [comment]
@tkurki I agree that all the mentioned points have been covered within this PR. And i also agree that changing the default behavior would be a problem

**Comment by godind** [comment]
Unless I made a mistake, I'm new at this GitHub stuff, I tested the PR in this thread by switching to the PR branch, build and ran the server.

I did a very long and detailed test and report to try and help.

**Pull request merged**